SecGPT - LlamaIndex Integration

Yuhao-W opened the pull request:
# Description

SecGPT is an LLM-based system that secures the execution of LLM apps via isolation. The key idea behind SecGPT is to isolate the execution of apps and to allow interaction between apps and the system only through well-defined interfaces with user permission. SecGPT can defend against multiple types of attacks, including app compromise, data stealing, inadvertent data exposure, and uncontrolled system alteration. We develop SecGPT using LlamaIndex because it supports several LLMs and apps and can be easily extended to include additional LLMs and apps. We implement SecGPT as a personal assistant chatbot, which the users can communicate with using text messages. 

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x ] I have performed a self-review of my own code
- [x ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x ] My changes generate no new warnings
- [x ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

nerdai reviewed (changes requested):
Hey @Yuhao-W,

I notice that we haven't used the standard process for creating a llama pack here. Could you follow the instructions at the link provided to get in the standard format? In particular, we use poetry for package dep manager as well as for building our python packages. For convenience we have a cli tool that helps you created these packs:

https://github.com/run-llama/llama_index/blob/main/CONTRIBUTING.md#2--contribute-a-pack-reader-tool-or-dataset-formerly-from-llama-hub

nerdai commented:
@Yuhao-W submitted a PR to your fork/main branch. It brings in the necessary pants build files to pass our checks.

[https://github.com/llm-platform-security/llama_index/pull/1](https://github.com/llm-platform-security/llama_index/pull/1)

Yuhao-W commented:
> @Yuhao-W looks like lint/fmt checks are failing. Can you please run:
> 
> `make lint` and `make format` then commit and push?

@nerdai Thanks, Andrei. I just fixed this.

Yuhao-W commented:
@nerdai Hi, Andrei. I see that some checks failed. Is there anything that needs to be changed?

Yuhao-W commented:
I think I figured out the errors and updated the package by mainly including dependency information in the pyproject.toml file under our package path. I also set up a unit test environment and ran it locally. The unit test was passed on my end. @nerdai  and/or @logan-markewich, would appreciate if you can review the changes!

nerdai commented:
> I think I figured out the errors and updated the package by mainly including dependency information in the pyproject.toml file under our package path. I also set up a unit test environment and ran it locally. The unit test was passed on my end. @nerdai and/or @logan-markewich, would appreciate if you can review the changes!

thanks, lets run the checks and see what happens!

Yuhao-W commented:
> > I think I figured out the errors and updated the package by mainly including dependency information in the pyproject.toml file under our package path. I also set up a unit test environment and ran it locally. The unit test was passed on my end. @nerdai and/or @logan-markewich, would appreciate if you can review the changes!
> 
> thanks, lets run the checks and see what happens!

Thanks @andrei, it failed again : ( this time because the `requirements.txt` was named `requirements.tx` 

I have made a new commit. Not sure but we may still see errors after, would appreciate a deeper look if it fails. Thank you!

nerdai commented:
@logan-markewich we're still running into some errors here. Perhaps we need to add a dependency in pants? This is the error we're seeing in the tests:

```
llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/sandbox.py:6: in <module>
    import tldextract
E   ModuleNotFoundError: No module named 'tldextract'
```

But `tldextract` is indeed included in the pyproject.toml as a dep for the project.

(CC @Yuhao-W)

nerdai commented:
@Yuhao-W got checks to pass 🥳. Needed to remove the requirements.txt file as it was tripping up pants having deps listed in both requirements.txt and pyproject.toml

nerdai commented on llama-index-packs/llama-index-packs-secgpt/examples/SecGPT.ipynb:
```diff
@@ -0,0 +1,741 @@
+{
```
<p>I'm a bit confused as to how these tools actually provide the fare price of both ride-sharing apps? Is this purely for illustration? In other words, is this notebook not actually functional?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/13127/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Yuhao-W replied:
You’re right! We developed these tools for attack illustration purposes. Developers can follow our examples to include their own real tools in SecGPT.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/examples/SecGPT.ipynb:
```diff
@@ -0,0 +1,741 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #4.    # A benign ride-sharing app - quick_ride
</pre><p>this one isn't benign is it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/13127/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Yuhao-W replied:
Yes, you are right. I have revised the description accordingly.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/examples/SecGPT.ipynb:
```diff
@@ -0,0 +1,741 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #20.    # A malicious ride-sharing app - metro hail
</pre><p>I think it was mentioned in the writeup before this cell block that QuickRide was the malicious app.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/13127/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Yuhao-W replied:
Same as above and revised accordingly.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/examples/SecGPT.ipynb:
```diff
@@ -0,0 +1,741 @@
+{
```
<p>Is there a way to show the case when not using SecGPT or having these measures turned off? In other words, can we see the case when the attack is successful?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/13127/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Yuhao-W replied:
That is a good idea! I have included another notebook, which defines a non-isolated LLM-based system to showcase the attack.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/planner.py:
```diff
@@ -0,0 +1,86 @@
+from llama_index.core import ChatPromptTemplate
+
+
+class HubPlanner:
```
Since there is a prompt here, I think we should subclass `PromptMixin`:

https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/prompts/mixin.py

Yuhao-W replied:
I have refined the prompt definition with a subclass of PromptMixin.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/planner.py:
```diff
@@ -0,0 +1,86 @@
+        lc_output_parser = JsonOutputParser()
+        self.output_parser = LangchainOutputParser(lc_output_parser)
+
+        self.query_engine = QueryPipeline(
```
this may be a bit of a name clash with llama-index ecosystem. As this is not really a `QueryEngine` but rather a `QueryPipeline`. If possible, would suggest using a different name: `query_pipeline`.

Yuhao-W replied:
I have revised accordingly.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/message.py:
```diff
@@ -0,0 +1,56 @@
+
+
+class Message:
+    def function_probe_request(self, spoke_id, function):
```
suggestion: maybe should this be a staticmethod?

nerdai replied:
similarly for all other funcs?

Yuhao-W replied:
I have revised accordingly.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/message.py:
Out of curiosity: have you considered using Pydantic BaseModel to represent Message? You can then subclass a BaseMessage. Pydantic can be helpful for validaton.

Yuhao-W replied:
That is a good idea. While I think the current representation of messages works, messages can also be represented using Pydantic BaseModel. However, as it requires non-trivial manual efforts, we can consider it in our  future revisions.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/spoke.py:
```diff
@@ -0,0 +1,112 @@
+            verbose=verbose,
+        )
+
+    def chat(
```
I think since this is a light wrapper on our React class, I think you can support stream_chat and its async versions as well.

Yuhao-W replied:
I have included other chatting methods accordingly.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/spoke.py:
```diff
@@ -0,0 +1,112 @@
+        self.spoke_operator.return_response(str(results))
+
+
+from llama_index.core.tools import FunctionTool
+
+
+def add_numbers(x: int, y: int) -> int:
+    """
+    Adds the two numbers together and returns the result.
+    """
+    return x + y
+
+
+if __name__ == "__main__":
+    llm = OpenAI(model="gpt-4-turbo", temperature=0.0, additional_kwargs={"seed": 0})
+    function_tool = FunctionTool.from_defaults(fn=add_numbers)
+    print(function_tool.metadata)
+    print(function_tool.metadata.get_parameters_dict())
+    spoke = Spoke(
+        tools=[function_tool],
+        collab_functions=["send_email", "draft_email", "read_email"],
+        llm=llm,
+        verbose=True,
+    )
+    spoke.chat("send a email to yuhao.wu@email.com, subject: hello, body: hello world")
```
This looks like it was used perhaps for testing while developing? I would suggest converting this into an acutal unit test and using mocking of LLMs.

For inspiration, see here: https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/agent/llama-index-agent-introspective/tests/test_self_reflection.py

Yuhao-W replied:
Thanks for pointing it out. he code is used for temporally testing. I have removed it.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/spoke_operator.py:
```diff
@@ -0,0 +1,70 @@
+    # Format and send the app request message to the hub
+    def make_request(self, functionality: str, request: dict):
+        # format the app request message
+        app_request_message = Message().app_request(
```
if these are staticmethods then you should be able to do `Message.app_request(...)` instead.

Yuhao-W replied:
I have revised accordingly.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/pyproject.toml:
```diff
@@ -0,0 +1,77 @@
+
+[tool.llamahub.class_authors]
+Hub = "Yuhao-W"
+HubOperator = "Yuhao-W"
+HubPlanner = "Yuhao-W"
+Message = "Yuhao-W"
+Socket = "Yuhao-W"
+Spoke = "Yuhao-W"
+SpokeOperator = "Yuhao-W"
+SpokeOutputParser = "Yuhao-W"
+TIMEOUT = "Yuhao-W"
+ToolImporter = "Yuhao-W"
+VanillaSpoke = "Yuhao-W"
+create_function_placeholder = "Yuhao-W"
+create_message_spoke_tool = "Yuhao-W"
+drop_perms = "Yuhao-W"
+get_user_consent = "Yuhao-W"
+set_mem_limit = "Yuhao-W"
```
all of these will show up in `llamahub.ai` which i don't think is ideal. Probably only makes sense to have the main one be discoverable i.e., `Hub`.

Yuhao-W replied:
It makes sense to me. I have revised.

nerdai commented on llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/hub.py:
```diff
@@ -0,0 +1,68 @@
+from .hub_operator import HubOperator
+
+
+class Hub(BaseLlamaPack):
```
The naming convention for packs is typically `XXXPack`. I would suggest Renaming this to:

`SecGPTPack` and then just create the alias for Hub i.e., add the below to the end of this file.

```python
Hub = SecGPTPack
```

Yuhao-W replied:
I have revised accordingly.

nerdai reviewed (commented):
@Yuhao-W Thanks for this contribution! I'm really excited about this :)

I left some comments on your PR. As another blanket comment, I do think your pack would greatly improve if you were able to include some doc/class strings throughout your code (i.e., quick descriptions of funcs/classes and its params/args).

Yuhao-W commented:
@nerdai Thanks for the feedback on the PR! I will address your comments and get back to you with an update soon.

nerdai commented:
hey @Yuhao-W: how's this coming along?

Yuhao-W commented:
> hey @Yuhao-W: how's this coming along?

Hi @nerdai . Thank you for checking with me. I’m currently working on it and will be able to make a push over the weekend.

Yuhao-W commented:
> @Yuhao-W Thanks for this contribution! I'm really excited about this :)
> 
> I left some comments on your PR. As another blanket comment, I do think your pack would greatly improve if you were able to include some doc/class strings throughout your code (i.e., quick descriptions of funcs/classes and its params/args).

Thanks for your suggestions. I have addressed all your comments and included doc/class strings for all classes and functions.

nerdai reviewed (approved):
Thanks @Yuhao-W -- this was a big one! Very excited to have this merged.
